packages/babel-register/src/loader.mjs

```diff
@@ -0,0 +1,287 @@
+// Many thanks to the author of https://gist.github.com/TravisMullen/16dfabca0db4bb4c32fb4f5ef297b25a
+
+/* eslint-disable import/extensions */
+
+// this is a workarround to make sure global._BABEL_ESM_REGISTER is set to true before anything else gets loaded
+import { readFile, access } from "node:fs/promises";
+import { constants } from "node:fs";
+import { dirname, extname, resolve as resolvePath } from "node:path";
+import { env } from "node:process";
+import { fileURLToPath, pathToFileURL } from "node:url";
+import Module from "module";
+
+import { addHook } from "pirates";
+import setup from "./hook-common.js";
+import { isInRegisterWorker } from "./is-in-register-worker.js";
+
+let JS_EXTENSIONS;
+let compile;
+let flag = {};
+let piratesRevert;
+
+global._BABEL_ESM_REGISTER = {
+  register(clientType, opts) {
+    if (!isInRegisterWorker) {
+      const { client, compile: _compile } = setup(clientType, opts);
+
+      flag = {};
+      compile = _compile;
+
+      // handle CJS
+      piratesRevert = addHook(compile, {
+        exts:
+          opts.extensions != null
+            ? opts.extensions
+            : client.getDefaultExtensions(),
+        ignoreNodeModules: false,
+      });
+
+      JS_EXTENSIONS = new Set(opts.extensions ?? client.getDefaultExtensions());
+    }
+  },
+  revert() {
+    if (piratesRevert) piratesRevert();
+    else return;
+    piratesRevert = null;
+    compile = null;
+    flag = {};
+  },
+};
+
+if (env.BABEL_REGISTER) {
+  const { default: opts } = await import(pathToFileURL(env.BABEL_REGISTER));
+  const { default: register } = await import("./experimental-worker.js");
+  register(opts);
+}
+
+function tryCJSResolve(specifier, context) {
+  const { parentURL } = context;
+  let fakeParent;
+
+  if (parentURL) {
+    const { pathname: parentPathname } = new URL(parentURL);
+
+    fakeParent = {
+      loaded: true,
+      filename: parentPathname,
+      id: parentPathname,
+      paths: Module._nodeModulePaths(dirname(parentPathname)),
+    };
+  }
+  let resolved;
+
+  try {
+    resolved = Module._resolveFilename(specifier, fakeParent);
+  } catch (error) {
+    // do nothing
+  }
+
+  if (resolved) {
+    return pathToFileURL(resolved).href;
+  }
+}
+
+async function tryESMResolve(specifier, context, nextResolve) {
+  try {
+    const next = await nextResolve(specifier, context);
+    return next.url;
+  } catch (error) {
+    // do nothing
+  }
+}
+
+async function exists(path) {
+  try {
+    await access(path, constants.R_OK);
+    return true;
+  } catch (error) {
+    // do nothing
+  }
+}
```

Comment:
https://github.com/nodejs/node/blob/e5c1fd7a2a1801fd75bdde23b260488e85453eb2/lib/fs.js#L290-L308

Looks like `existsSync` is better.

Reply:
+1 for `existsSync`. Node.js had an asynchronus `exist` API, that had been deprecated because checking if a file exists was effectively a sync operation anyway.